{% extends "base.html" %}

{% block content %}
    <div id="page-wrapper">
        <div class="row">
            <div class="col-lg-12">
                <h1 class="page-header">sRNAtoolbox: a collection of small RNA analysis tools</h1>
            </div>
            <!-- /.col-lg-12 -->
        </div>
    <!-- /.row -->
    <div class="row">
    <div class="col-lg-12">
    <div class="row row-grid">
        <div class="well well-lg col-lg-12">
            <h4>sRNAtoolbox is aimed to provide small RNA researchers with several useful tools including sRNA
                expression profiling from deep sequencing experiments and several downstream analysis tools.
                The center piece of sRNAtoolbox is sRNAbench, which allows for expression profiling
{#                and prediction of novel microRNAs #}
                from deep sequencing experiments.
                The other tools can be either launched on sRNAbench results, or independently using the
                appropriate file formats. For more information please check our publications on
                <a href="https://doi.org/10.1093/nar/gkac363"><i>Nucleic Acids Research</i></a> (and
                <a href="https://doi.org/10.1093/nar/gkz415"><i>previous version</i></a>)
            </h4>



            <br>
            <h4>Currently the toolbox comprises the following methods:</h4>


        </div>
    </div>

    <div class="row row-grid">


    <div class="panel panel-primary">
        <div class="panel-heading">
            <h3>Methods</h3>
        </div>
        <div class="panel-body">

            <!-- Nav tabs -->
            <ul class="nav nav-tabs">
                <li class="active"><a href="#sRNAbench" data-toggle="tab">sRNAbench</a>
                </li>
{#                <li class=""><a href="#batch" data-toggle="tab">sRNAbench (batch mode)</a>#}
{#                </li>#}
                <li class=""><a href="#sRNAde" data-toggle="tab">sRNAde</a>
                </li>
                <li class=""><a href="#sRNAblast" data-toggle="tab">sRNAblast</a>
                </li>
                <li class=""><a href="#miRNAconsTarget" data-toggle="tab">miRNAconsTarget</a>
                </li>
                <li class=""><a href="#sRNAcons" data-toggle="tab">sRNAcons</a>
                </li>
{#                <li class=""><a href="#sRNAjBrowser" data-toggle="tab">sRNAjBrowser</a>#}
{#                </li>#}
{#                <li class=""><a href="#sRNAjBrowserDE" data-toggle="tab">sRNAjBrowserDE</a>#}
{#                </li>#}
{#                <li class=""><a href="#sRNAfuncTerms" data-toggle="tab">sRNAfuncTerms</a>#}
{#                </li>#}
            </ul>

            <!-- Tab panes -->

            <div class="tab-content">
                <div class="tab-pane fade active in" id="sRNAbench">
                    <a href="{{ settings.SUB_SITE }}/newbench/"> <h3>sRNAbench:</h3> <h4>Expression profiling of small RNAs and prediction of novel microRNAs from deep sequencing data</h4>  </a>
                    <h4 style="line-height: 150%">sRNAbench is mainly designed to be uses in model species, or at least species with
                        an available genome assembly or fasta-formatted annotations. Several commonly used input formats like fastq, sra, fasta or read/count are supported.
                        It performs expression profiling of known microRNAs and other types of small RNAs, detection and analysis of isomiRs,
                        outputs several graphical summaries and predicts novel microRNAs both in plants and animals.</h4>
                </div>

{#                <div class="tab-pane fade" id="batch">#}
{#                    <a href="{{ settings.SUB_SITE }}/newbench/"> <h3>sRNAbench (batch mode):</h3> <h4> Parallel expression profiling of small RNAs for several files at once</h4>  </a>#}
{#                    <h4 style="line-height: 150%"> Batch mode is a new feature of sRNAbench that allows for intuitive profiling of several samples at once.#}
{#                    If you have several samples with the same sequencing approach (library preparation) this tool will profile all these samples using the same#}
{#                    bioinformatics protocol. This will also guarantee that all your samples are profiled following the exact same steps.</h4>#}
{#                </div>#}

                <div class="tab-pane fade" id="sRNAde">
                    <a href="{{ settings.SUB_SITE }}/srnade/"> <h3>sRNAde:</h3> <h4>Differential expression analysis </h4></a>
                    <h4 style="line-height: 150%">sRNAde is a module for the detection of differentially expressed sRNA. The input is
                                either a number of sRNAbench output folders or a user-given expression matrix.
                                For the detection of differentially expressed RNAs it applies 5 widely used methods:
                                t-test on Reas per Million, edgeR, DEseq, DEseq2 and noiseq. Besides individual methods output, it provides a
                                consensus differential expression file. Additionally, this tool can perform cluster
                                analysis (heatmaps), isomiR analysis and it gives a summary on the sequencing statistics
                                of all used samples (if the input has been from sRNAbench).</h4>
                </div>

                <div class="tab-pane fade" id="sRNAgFree">
                    <a href="/srnagfree/" > <h3>sRNAgFree:</h3>  <h4>Profiling and detection of novel microRNAs in non-model specie</h4>  </a>
                    <h4 style="line-height: 150%">Currently, microRNA deep sequencing data is available for many species that do not count
                        with a proper genome assembly. In such cases, frequently simple homology based detection
                        methods are applied. These methods can lead to false positive predictions but will also
                        fail to detect microRNAs that are not present in the reference database (like miRBase).
                        sRNAgFree is aimed for the prediction and expression profiling of microRNAs in
                        non-model species. It based on the duplex properties formed by the guide and passenger
                        strand and applies homology and learning form known binding and Drosha/Dicer processing
                        patterns.</h4>
                </div>

                <div class="tab-pane fade" id="sRNAblast">
                    <a href="{{ settings.SUB_SITE }}/srnablast/"> <h3>sRNAblast:</h3> <h4>Blast analysis of deep sequencing reads against a local nt/nr (NCBI link) database</h4>  </a>
{#                    <a href="{{ settings.SUB_SITE }}/srnablast"> <h3>sRNAblast:</h3> <h4>Blast analysis of deep sequencing reads against a local nt/nr (NCBI link) database</h4>  </a>#}
                    <h4 style="line-height: 150%">This tool is principally aimed for the analysis of reads that could not be mapped in
                                sRNAbench or other profiling tools. The results could either point towards contamination
                                sources or biological meaningful information like the presence of unexpected viral or
                                bacterial RNA molecules.</h4>
                </div>


                <div class="tab-pane fade" id="miRNAconsTarget">
                    <a href="{{ settings.SUB_SITE }}/amirconstarget/"> <h3>animalconsTarget:</h3> <h4>Predict microRNA targets on user defined sets of microRNAs and 3'UTRs </h4> </a>
{#                    <a href="{{ settings.SUB_SITE }}/amirconstarget"> <h3>animalconsTarget:</h3> <h4>Predict microRNA targets on user defined sets of microRNAs and 3'UTRs </h4> </a>#}
                    <h4 style="line-height: 150%">Most microRNA target prediction tools are implemented as webservers on a limited number of
                                species, or do allow only the prediction of a limited number of microRNAs and/or mRNAs.
                                miRNAconsTargets uses a selection microRNA target prediction programs and
                                applies them to the user supplied sets of microRNAs and 3'UTRs. It reports both,
                                individual predictions and a consensus prediction. </h4>
                    <a href="{{ settings.SUB_SITE }}/pmirconstarget/"> <h3>plantconsTarget:</h3> <h4>Predict microRNA targets on user defined sets of microRNAs and 3'UTRs or CDS sequences </h4> </a>
{#                    <a href="{{ settings.SUB_SITE }}/pmirconstarget"> <h3>plantconsTarget:</h3> <h4>Predict microRNA targets on user defined sets of microRNAs and 3'UTRs or CDS sequences </h4> </a>#}
                    <h4 style="line-height: 150%">Most microRNA target prediction tools are implemented as webservers on a limited number of
                                species, or do allow only the prediction of a limited number of microRNAs and/or mRNAs.
                                miRNAconsTargets uses a selection microRNA target prediction programs and
                                applies them to the user supplied sets of microRNAs and 3'UTRs or CDS sequences. It reports both,
                                individual predictions and a consensus prediction. </h4>
                </div>

                <div class="tab-pane fade" id="sRNAcons">
                    <a href="{{ settings.SUB_SITE }}/srnacons/" > <h3>sRNAcons:</h3> <h4>Estimation of the conservation depth of small RNAs</h4>  </a>
                    <h4 style="line-height: 150%"> User provided small RNA sequences (fasta format) are mapped against all assemblies available
                        in sRNAtoolbox. The tool has two different outputs, i) The conservation depth for all small RNA input sequences, i.e
                        the percentage of genomes in which the sequence was found, and ii)the percentage of mapped input sequences per genome.</h4>
                </div>

                <div class="tab-pane fade" id="sRNAjBrowser">
                    <a href="/srnajbrowser/" > <h3>sRNAjBrowser:</h3> <h4>Browse the small RNA expression values in a genome context </h4> </a>
                    <h4 style="line-height: 150%">In several small RNA research fields, the visual exploration of small RNA expression
                                pattern in a genome context will be helpful. sRNAjBrowser allows the user to analyse
                                the expression data in a genome context by means of a jBrowser implementation.
                                Furthermore, this tool is connected to NGSmethDB and can therefore also display
                                some selected methylation tracks.</h4>
                </div>

                <div class="tab-pane fade" id="sRNAjBrowserDE">
                    <a href="/srnajbrowserde/" > <h3>sRNAjBrowserDE:</h3> <h4>Visualize differential expression as a function of read length </h4> </a>

                        <h4 style="line-height: 150%">
                        sRNAjBrowserDE extends on sRNAjBrowser as it: i) allows the comparison of two experimental
                            groups, i.e. it visualizes the differential expression in a genome context and ii) it
                            allows to visualize the expression as a function of read length. This might me particularly
                            interesting for plant research as 24nt long and 21/22nt long reads have very different functions. </h4>
                </div>

                <div class="tab-pane fade" id="sRNAfuncTerms">
                    <a href="/srnafuncterms/" > <h3>sRNAfuncTerms:</h3> <h4>Determine over represented functional annotations in target gene set</h4>  </a>
                    <h4 style="line-height: 150%">Overrepresented functional annotations in a list of target genes might give useful hints
                                on the functional implications of the microRNAs. sRNAfuncTerms takes a set of microRNAs
                                as input, retrieves the target genes from the underlying database and detected
                                overrepresented GO-terms among the target genes. Apart from the complete set of microRNAs,
                                it tests also all microRNA modules (combinations) which allow the user to detect the microRNAs
                                that act on the same pathways or share the same functions.</h4>
                </div>

            </div>
        </div>
        </div>

{#        <div class="panel panel-green">#}
{#        <div class="panel-heading">#}
{#            <h3>Workflows</h3>#}
{#        </div>#}
{#        <div class="panel-body">#}
{##}
{#            <!-- Nav tabs -->#}
{#            <ul class="nav nav-tabs">#}
{#                <li class="active"><a href="#miRNAfuncTargets" data-toggle="tab">miRNAfuncTargets</a>#}
{#                </li>#}
{#            </ul>#}
{##}
{#            <!-- Tab panes -->#}
{##}
{#            <div class="tab-content">#}
{#                <div class="tab-pane fade active in" id="miRNAfuncTargets">#}
{#                    <a href=""> <h3>miRNAfuncTargets:</h3> <h4>Cross-species microRNA target prediction and enrichment analysis of functional annotations. (Currently not available)</h4>  </a>#}
{#                    <a href="{{ settings.SUB_SITE }}/mirnafunctargets/"> <h3>miRNAfuncTargets:</h3> <h4>Cross-species microRNA target prediction and enrichment analysis of functional annotations.</h4>  </a>#}
{#                    <h4 style="line-height: 150%">This workflow connects directly sRNAconsTargets and sRNAfuncTerms.#}
{#                        It allows the user to i) predict the putative target genes of a set of exogenous microRNAs and#}
{#                        ii) determine the enriched functional annotations, i.e. which pathways or functions are enriched#}
{#                        for the target genes.</h4>#}
{#                </div>#}
{##}
{#        </div>#}
{#            </div>#}
{#            </div>#}

        <div class="panel panel-info">
        <div class="panel-heading">
            <h3>Helper Tools</h3>
        </div>
        <div class="panel-body">

            <h4 style="line-height: 150%">These tools are intended to help the user to either setting up a local sRNAbench database
                (Ensembl Parser, NCBI Parser, RNA Central Parser, genomic tRNA Parser) or to prepare input
                data for other sRNAtoolbox tools. If you use any of the generated data in your publication, please
                cite the papers given at the data retrieval page (NCBI, ENSEMBL, genomic tRNA and RNA central databases).</h4>
<br><h4>Tools: </h4>
            <ul class="list-inline">
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/ensemlparser/">Ensembl Parser</a>
                                </li>
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/ncbiparser/">NCBI Parser</a>
                                </li>
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/rnacentralparser/">RNA central parser</a>
                                </li>
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/trnaparser/">Genomic tRNA database parser</a>
                                </li>
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/removedup/">Remove duplicates from a fasta file</a>
                                </li>
                                <li>
                                    <a href="{{ settings.SUB_SITE }}/helper/extract/">Extract Sequences from a fasta file</a>
                                </li>
                            </ul>
        </div>



		</div>



            </div>

    <script type="text/javascript">
    function showValue(newValue)
    {
        document.getElementById("range").innerHTML=newValue;
    }
    </script>

    <!-- /.row -->
        </div>
        <!-- /#page-wrapper -->
{% endblock %}


